### PR TITLE
Plugins: Refactor creation of plugin context to dedicated service

### PR DESCRIPTION
**Pull request by wbrowne**
**What is this feature?**

Migrate all plugin client consumers to use the same service to create the outgoing plugin context

**Why do we need this feature?**

Consistency 

**Who is this feature for?**

Plugins platform

**Special notes for your reviewer:**
- Can try split this to be smaller if necessary (note that a chunk of these changes are tests+mocks)
- There will be a follow up to create an interface for the plugin context service so that Enterprise can have its own implementation

Please check that:
- [X] It works as expected from a user's perspective.
- [X] If this is a pre-GA feature, it is behind a feature toggle.
- [X] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/expr/transform.go [hunk 1]
```diff
@@ -151,9 +152,3 @@ func hiddenRefIDs(queries []Query) (map[string]struct{}, error) {
 	}
 	return hidden, nil
 }
-
-func (s *Service) decryptSecureJsonDataFn(ctx context.Context) func(ds *datasources.DataSource) (map[string]string, error) {
```
Comment (andresmgot) [inline pkg/expr/transform.go:155]: these changes seem unrelated?

File: pkg/services/pluginsintegration/plugincontext/key.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+	AppKey(ctx context.Context, pluginID string, orgID int64) string
+}
+
+func ProvideKeyService() *KeyService {
+	return &KeyService{}
+}
+
+type KeyService struct{}
+
+func (s *KeyService) DataSourceKey(_ context.Context, dsSettings *backend.DataSourceInstanceSettings) string {
```
Comment (andresmgot) [inline pkg/services/pluginsintegration/plugincontext/key.go:21]: `DataSourceID`?  or is this meant to return something different? Some documentation would help

File: pkg/services/pluginsintegration/plugincontext/plugincontext.go [hunk 3]
```diff
@@ -11,68 +11,83 @@ import (
...
+		User:     adapters.BackendUserFromSignedInUser(user),
+	}
+
+	if plugin.IsApp() {
+		appSettings, err := p.appInstanceSettings(ctx, pluginID, user)
+		if err != nil {
+			return backend.PluginContext{}, err
+		}
+		pCtx.AppInstanceSettings = appSettings
+		//pCtx.Key = p.keyProvider.AppKey(ctx, pluginID, user.OrgID)
```
Comment (andresmgot) [inline pkg/services/pluginsintegration/plugincontext/plugincontext.go:63]: leftover? I don't see the `keyProvider` being used. Am I missing something?

File: pkg/services/pluginsintegration/plugincontext/plugincontext.go [hunk 4]
```diff
@@ -11,68 +11,83 @@ import (
...
+		return backend.PluginContext{}, err
 	}
 
 	datasourceSettings, err := adapters.ModelToInstanceSettings(ds, p.decryptSecureJsonDataFn(ctx))
 	if err != nil {
-		return pCtx, exists, fmt.Errorf("%v: %w", "Failed to convert datasource", err)
+		return pCtx, err
 	}
 	pCtx.DataSourceInstanceSettings = datasourceSettings
+	//pCtx.Key = p.keyProvider.DataSourceKey(ctx, datasourceSettings)
```
Comment (andresmgot) [inline pkg/services/pluginsintegration/plugincontext/plugincontext.go:82]: same here?

**Review by andresmgot — COMMENTED** [review]
Mostly looks good to me but the `KeyService` seems to not be used, is that expected?

File: pkg/services/pluginsintegration/plugincontext/plugincontext.go (hunk 4 above)
Comment (wbrowne) [reply pkg/services/pluginsintegration/plugincontext/plugincontext.go:82]: Yeah sorry I should mention that changes are required to the SDK to introduce a new field (for example `Key` to the plugin context) - https://github.com/grafana/grafana-plugin-sdk-go/blob/main/backend/common.go#L128. This is purely to demonstrate what it might look like

File: pkg/services/pluginsintegration/plugincontext/plugincontext.go (hunk 3 above)
Comment (wbrowne) [reply pkg/services/pluginsintegration/plugincontext/plugincontext.go:63]: Ditto https://github.com/grafana/grafana/pull/66451#discussion_r1165662601

File: pkg/expr/transform.go (hunk 1 above)
Comment (wbrowne) [reply pkg/expr/transform.go:155]: Because `dataSourceService` was removed as a dependency of `Service`, now the pluginContextProvider will take care of this instead

File: pkg/services/pluginsintegration/plugincontext/key.go (hunk 2 above)
Comment (wbrowne) [reply pkg/services/pluginsintegration/plugincontext/key.go:21]: Yeah sorry without docs or the right context this can be confusing. 

So right now in the SDK, when we cache either datasource or app "instances", the cache key is defined in the SDK:
- For [apps](https://github.com/grafana/grafana-plugin-sdk-go/blob/main/backend/app/instance_provider.go#L51)
- For [data sources](https://github.com/grafana/grafana-plugin-sdk-go/blob/main/backend/datasource/instance_provider.go#L48)

A new `pluginContext.Key` field can be used to replace this in the SDK so that Grafana can decide how the instances are cached. Since there is a separate key for Apps and DataSources, I'm keeping the same behaviour here by using this `KeyProvider` to be used as part of `plugincontext.Provider`. Does that make sense? 🤔

Comment (andresmgot) [reply pkg/services/pluginsintegration/plugincontext/key.go:21]: yes :+1: okay that makes sense. The only thing to have into account is plugins using an old Grafana version with a new SDK version (but we can add a fallback path if `pluginContext.Key` is empty).

File: pkg/services/pluginsintegration/plugincontext/plugincontext.go (hunk 4 above)
Comment (andresmgot) [reply pkg/services/pluginsintegration/plugincontext/plugincontext.go:82]: okay, so what this PR is doing, if I understand that correctly is:

 - Simplify the usage of pluginContext
 - Define a KeyService
 - (To be enabled) Enable the KeyService within the pluginContext

That makes sense :+1: Maybe the only comment I would do is to use `CacheKeyService` for the new service, so it's clear what is its purpose.

**Review by andresmgot — APPROVED** [review]
LGTM

File: pkg/services/pluginsintegration/plugincontext/plugincontext.go [hunk 5]
```diff
@@ -82,27 +92,26 @@ func (p *Provider) pluginContext(ctx context.Context, pluginID string, user *use
 		// pluginsettings.ErrPluginSettingNotFound is expected if there's no row found for plugin setting in database (if non-app plugin).
 		// If it's not this expected error something is wrong with cache or database and we return the error to the client.
 		if !errors.Is(err, pluginsettings.ErrPluginSettingNotFound) {
-			return backend.PluginContext{}, false, fmt.Errorf("%v: %w", "Failed to get plugin settings", err)
+			return &backend.AppInstanceSettings{}, fmt.Errorf("%v: %w", "Failed to get plugin settings", err)
```
Comment (marefr) [inline pkg/services/pluginsintegration/plugincontext/plugincontext.go:95]: Nit. Can return nil instead

File: pkg/services/pluginsintegration/plugincontext/plugincontext.go [hunk 6]
```diff
@@ -82,27 +92,26 @@ func (p *Provider) pluginContext(ctx context.Context, pluginID string, user *use
...
 		}
 	} else {
 		jsonData, err = json.Marshal(ps.JSONData)
 		if err != nil {
-			return backend.PluginContext{}, false, fmt.Errorf("%v: %w", "Failed to unmarshal plugin json data", err)
+			return &backend.AppInstanceSettings{}, fmt.Errorf("%v: %w", "Failed to unmarshal plugin json data", err)
```
Comment (marefr) [inline pkg/services/pluginsintegration/plugincontext/plugincontext.go:100]: Nit. Can return nil instead

**Review by marefr — APPROVED** [review]
LGTM great work. Added a few nits

File: pkg/services/ngalert/eval/eval.go [hunk 7]
```diff
@@ -247,6 +247,7 @@ func getExprRequest(ctx EvaluationContext, data []models.AlertQuery, dsCacheServ
 	req := &expr.Request{
 		OrgId:   ctx.User.OrgID,
 		Headers: buildDatasourceHeaders(ctx.Ctx),
+		User:    ctx.User,
```
Comment (yuri-tceretian) [inline pkg/services/ngalert/eval/eval.go:250]: How do you use this User downstream? Asking because the scheduler that executes this method uses a fake user 
https://github.com/grafana/grafana/blob/9eb10bee1f2d249b746c8011f7446b10eec94c98/pkg/services/ngalert/schedule/schedule.go#L574-L589

Comment (wbrowne) [reply pkg/services/ngalert/eval/eval.go:250]: I have seen it only used as purely informational - and the plugin SDK docs highlights this detail about it potentially coming from an alerting context https://github.com/grafana/grafana-plugin-sdk-go/blob/main/backend/common.go#L137-L141 and thus not to be relied on.

File: pkg/expr/nodes.go [hunk 8]
```diff
@@ -207,20 +206,14 @@ func (dn *DSNode) Execute(ctx context.Context, now time.Time, _ mathexp.Vars, s
 	ctx, span := s.tracer.Start(ctx, "SSE.ExecuteDatasourceQuery")
 	defer span.End()
 
-	dsInstanceSettings, err := adapters.ModelToInstanceSettings(dn.datasource, s.decryptSecureJsonDataFn(ctx))
+	pCtx, err := s.pCtxProvider.GetWithDataSource(ctx, dn.datasource.Type, dn.request.User, dn.datasource)
```
[the remark this answers is not in the document]
Comment (wbrowne) [reply pkg/expr/nodes.go:209]: @kylebrandt This 🤞 should be fixed by now by https://github.com/grafana/grafana/pull/69811 - unless you see something else?

[the remark this answers is not in the document]
Comment (kylebrandt) [reply pkg/expr/nodes.go:209]: Yes that fixes it thank you. My timing was just terrible ... started to dive in RQ code and using enterprise right before the fix you mention landed :-/
